File: sportify_mobile/lib/services/chat_service.dart
import 'package:http/http.dart' as http;
import 'package:stomp_dart_client/stomp_dart_client.dart';
import 'dart:convert';
import '../utils/constants.dart';

class ChatService {
  Future<List<Map<String, dynamic>>> getChatRooms(String token) async {
    final response = await http.get(
      Uri.parse('$BASE_URL$GET_CHATROOMS'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode == 200) {
      List<dynamic> jsonResponse = jsonDecode(response.body);
      return jsonResponse.cast<Map<String, dynamic>>();
    } else {
      throw Exception('Failed to load chat rooms');
    }
  }

//CHAT REALTIME
  StompClient? stompClient;
  Function(Map<String, dynamic>)? onMessageReceived;
  Function(List<Map<String, dynamic>>)? onChatRoomsUpdated;
  ChatService({this.onMessageReceived, this.onChatRoomsUpdated});

  /// Kết nối WebSocket (dành cho MessageScreen)
  void connectForChatRooms(String token, int userId) {
    stompClient = StompClient(
      config: StompConfig.sockJS(
        url: '$BASE_URL/ws',
        onConnect: (StompFrame frame) {
          print('WebSocket connected for chat rooms!');
          subscribeToChatRooms(userId);
        },
        beforeConnect: () async {
          print('Connecting to WebSocket for chat rooms...');
        },
        onWebSocketError: (dynamic error) => print('WebSocket Error: $error'),
      ),
    );
    stompClient?.activate();
  }

  /// Đăng ký lắng nghe danh sách phòng chat cập nhật
  void subscribeToChatRooms(int userId) {
    print('subscribe to chatrooms with userId:  $userId');
    stompClient?.subscribe(
      destination: '/topic/chatrooms/$userId',
      callback: (StompFrame frame) {
        if (frame.body != null) {
          final List<dynamic> updatedChatRooms = jsonDecode(frame.body!);
          // 👉 Log ra thông báo khi danh sách phòng thay đổi
          print('🔔 Danh sách phòng chat đã thay đổi! Số phòng: ');

          if (updatedChatRooms.isNotEmpty) {
            print('Rooms updated');
            final firstRoom = updatedChatRooms[0];
            print('Phòng đầu tiên: ${firstRoom['roomId']}');
            print('Phòng đầu tiên: ${firstRoom['lastMessage']}');
          } else {
            print('⚠️ Không có phòng nào trong danh sách.');
          }
          if (onChatRoomsUpdated != null) {
            onChatRoomsUpdated!(updatedChatRooms.cast<Map<String, dynamic>>());
          }
        }
      },
    );
  }

  /// Kết nối WebSocket (dành cho ChatScreen)
  void connectForMessages(String token, int roomId) {
    stompClient = StompClient(
      config: StompConfig.sockJS(
        url: '$BASE_URL/ws',
        onConnect: (StompFrame frame) {
          print('WebSocket connected for messages!');
          subscribeToMessages(roomId);
        },
        beforeConnect: () async {
          print('Connecting to WebSocket for messages...');
        },
        onWebSocketError: (dynamic error) => print('WebSocket Error: $error'),
      ),
    );
    stompClient?.activate();
  }

  /// Đăng ký lắng nghe tin nhắn mới từ WebSocket
  void subscribeToMessages(int roomId) {
    stompClient?.subscribe(
      destination: '/topic/messages/$roomId',
      callback: (StompFrame frame) {
        if (frame.body != null) {
          final Map<String, dynamic> message = jsonDecode(frame.body!);
          if (onMessageReceived != null) {
            onMessageReceived!(message);
          }
        }
      },
    );
  }

  /// Gửi tin nhắn qua WebSocket
  void sendMessage(int roomId, int senderId, String content) {
    if (stompClient != null && stompClient!.connected) {
      final message = {
        'senderId': senderId,
        'content': content,
      };
      stompClient!.send(
        destination: '/app/chat/$roomId',
        body: jsonEncode(message),
      );
    } else {
      print('WebSocket not connected!');
    }
  }

  /// Lấy lịch sử tin nhắn từ API
  Future<List<Map<String, dynamic>>> getMessages(
      String token, int roomId) async {
    final response = await http.get(
      Uri.parse('$BASE_URL/api/messages/history/$roomId'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode == 200) {
      List<dynamic> jsonResponse = jsonDecode(response.body);
      return jsonResponse.cast<Map<String, dynamic>>();
    } else {
      throw Exception('Failed to load messages');
    }
  }

  /// Đóng kết nối WebSocket
  void disconnect() {
    stompClient?.deactivate();
  }

  /// Đánh dấu đã đọc toàn bộ tin nhắn trong phòng chat
  Future<void> markAllMessagesAsRead(String token, int chatRoomId) async {
    final url = Uri.parse('$BASE_URL/api/messages/mark-read/$chatRoomId');

    try {
      final response = await http.put(
        url,
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        print('Đã đánh dấu đã đọc : $chatRoomId');
      } else {
        print('Lỗi khi đánh dấu đã đọc: ${response.statusCode}');
      }
    } catch (e) {
      print('Lỗi kết nối API đánh dấu đã đọc: $e');
    }
  }
}

File: FE_User/src/services/chatService.js
// import axios from "axios";
// import { Client } from "@stomp/stompjs";
// import SockJS from "sockjs-client";

// const baseURL = process.env.REACT_APP_BASE_URL.replace(/\/api\/?$/, "");
// const API_URL = `${baseURL}/api/chat`; // Cập nhật API URL
// const WEBSOCKET_URL = `${baseURL}/ws`;

// let stompClient = null;
// let messageCallback = null;

// // 🟢 Kết nối WebSocket
// const connectWebSocket = (userId, callback) => {
//   console.log("Connect to: ", API_URL);
//   if (stompClient && stompClient.connected) return;

//   const socket = new SockJS(WEBSOCKET_URL);
//   stompClient = new Client({
//     webSocketFactory: () => socket,
//     reconnectDelay: 5000,
//     onConnect: () => {
//       console.log("🔗 WebSocket Connected");
//       messageCallback = callback;

//       // Lắng nghe danh sách phòng chat của user
//       stompClient.subscribe(`/topic/chatrooms/${userId}`, (message) => {
//         const updatedChatRooms = JSON.parse(message.body);
//         if (messageCallback) messageCallback(updatedChatRooms);
//       });
//     },
//     onDisconnect: () => console.log("❌ WebSocket Disconnected"),
//     onStompError: (error) => console.error("WebSocket Error:", error),
//   });

//   stompClient.activate();
// };

// // 🛑 Ngắt kết nối WebSocket
// const disconnectWebSocket = () => {
//   if (stompClient) {
//     stompClient.deactivate();
//     console.log("🔌 WebSocket Disconnected");
//   }
// };

// // 📨 Lấy danh sách phòng chat
// const getChatRooms = async () => {
//   const token = localStorage.getItem("token-login");
//   const response = await axios.get(`${API_URL}/rooms`, {
//     headers: { Authorization: `Bearer ${token}` },
//   });
//   console.log(response.data);
//   return response.data;
// };

// const chatService = {
//   getChatRooms,
//   connectWebSocket,
//   disconnectWebSocket,
// };

// export default chatService;

import axios from "axios";
import { Client } from "@stomp/stompjs";
import SockJS from "sockjs-client";
import { infoUser } from "./infoUser";

const baseURL = process.env.REACT_APP_BASE_URL.replace(/\/api\/?$/, "");
const API_URL = `${baseURL}/api/chat`; // API URL
const WEBSOCKET_URL = `${baseURL}/ws`; // WebSocket URL

let stompClient = null;
let chatRoomsCallback = null;
let messageCallbacks = {};

// 🟢 Kết nối WebSocket & Subscribe danh sách phòng chat
const connectWebSocket = (token, userId, callback) => {
  if (!token) {
    console.error("🚨 Không có token, không thể kết nối WebSocket!");
    return;
  }
  if (!userId) {
    console.error("🚨 Không có userId, không thể kết nối WebSocket!");
    return;
  }
  if (stompClient && stompClient.connected) return;

  const socket = new SockJS(WEBSOCKET_URL);
  stompClient = new Client({
    webSocketFactory: () => socket,
    reconnectDelay: 5000,
    onConnect: () => {
      console.log(" WebSocket Connected");
      chatRoomsCallback = callback;

      // 📌 Subscribe danh sách phòng chat
      stompClient.subscribe(`/topic/chatrooms/${userId}`, (message) => {
        const updatedChatRooms = JSON.parse(message.body);
        console.log(" Nhận danh sách phòng chat từ server:", updatedChatRooms);
        if (chatRoomsCallback) chatRoomsCallback(updatedChatRooms);
      });
    },
    onDisconnect: () => console.log("❌ WebSocket Disconnected"),
    onStompError: (error) => console.error("WebSocket Error:", error),
  });

  stompClient.activate();
};

// 🛑 Ngắt kết nối WebSocket
const disconnectWebSocket = () => {
  if (stompClient) {
    stompClient.deactivate();
    console.log("🔌 WebSocket Disconnected");
  }
};

// 🔔 Subscribe tin nhắn của phòng chat
const subscribeToRoomMessages = (roomId, callback) => {
  if (!stompClient || !stompClient.connected) return;

  console.log(`📩 Subscribing to messages of room ${roomId}`);
  if (messageCallbacks[roomId]) return; // Nếu đã sub thì không sub lại

  const subscription = stompClient.subscribe(
    `/topic/messages/${roomId}`,
    (message) => {
      const newMessage = JSON.parse(message.body);
      callback(newMessage);
    }
  );

  messageCallbacks[roomId] = subscription;
};

// ❌ Unsubscribe tin nhắn của phòng chat
const unsubscribeFromRoomMessages = (roomId) => {
  if (messageCallbacks[roomId]) {
    messageCallbacks[roomId].unsubscribe();
    delete messageCallbacks[roomId];
    console.log(`🚫 Unsubscribed from messages of room ${roomId}`);
  }
};

// 📨 Lấy danh sách phòng chat
const getChatRooms = async () => {
  const token = localStorage.getItem("token-login");
  const response = await axios.get(`${API_URL}/rooms`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.data;
};

const getChatHistory = async (chatRoomId) => {
  const token = localStorage.getItem("token-login");
  const response = await axios.get(
    `${baseURL}/api/messages/history/${chatRoomId}`,
    {
      headers: { Authorization: `Bearer ${token}` },
    }
  );
  return response.data;
};

const sendMessage = (chatRoomId, senderId, message) => {
  if (!stompClient || !stompClient.connected) {
    console.error("❌ WebSocket chưa kết nối!");
    return;
  }

  const messageData = {
    senderId: senderId,
    content: message,
  };

  stompClient.publish({
    destination: `/app/chat/${chatRoomId}`,
    body: JSON.stringify(messageData),
  });

  console.log("📨 Đã gửi tin nhắn:", messageData);
};

const chatService = {
  getChatRooms,
  getChatHistory,
  connectWebSocket,
  disconnectWebSocket,
  subscribeToRoomMessages,
  unsubscribeFromRoomMessages,
  sendMessage,
};

export default chatService;
